File: projects/order-creation/core/src/main/java/com/robinfood/core/dtos/transactionrequestdto/OrderDTO.java
package com.robinfood.core.dtos.transactionrequestdto;

import com.robinfood.core.dtos.DeductionDTO;
import com.robinfood.core.dtos.transactionresponsedto.OrderDiscountResponseDTO;
import com.robinfood.core.dtos.transactionresponsedto.OrderStatusResponseDTO;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import javax.validation.Valid;
import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Positive;
import javax.validation.constraints.Size;
import java.io.Serializable;
import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class OrderDTO implements Serializable {

    private static final long serialVersionUID = 6623614449999577563L;

    @Positive
    private Long billingResolutionId;

    @NotNull
    private FinalProductBrandDTO brand;

    private Map<String, Object> criteriaInfo;

    @Builder.Default
    private List<CouponDTO> coupons = new ArrayList<>();

    private Map<String, Object> couponCriteriaInfo;

    private BigDecimal consumptionValue = BigDecimal.ZERO;

    @NotNull
    @Positive
    private Long deliveryTypeId;

    @Builder.Default
    private List<DeductionDTO> deductions = new ArrayList<>();

    @NotNull
    @Size
    @Valid
    private List<DiscountDTO> discounts;

    private List<OrderDiscountResponseDTO> discountsApplied;

    @NotNull
    @Valid
    @Size(min = 1)
    private List<FinalProductDTO> finalProducts;

    @Valid
    private FlagDTO flags;

    private BigDecimal foodcoins;

    @Positive
    private Long id;

    private String invoiceNumber;

    @Builder.Default
    private Boolean isMultiBrand = Boolean.FALSE;

    private Boolean hasConsumption = Boolean.FALSE;

    @NotBlank
    private String notes;

    private String orderNumber;

    @NotNull
    @Valid
    private OriginDTO origin;

    @NotNull
    @Positive
    private Long paymentModelId;

    private PosResolutionDTO posResolution;

    @NotNull
    @Size
    @Valid
    private List<ServiceDTO> services;

    private BigDecimal subtotal;

    private OrderStatusResponseDTO status;

    @NotNull
    @Valid
    private StoreDTO store;

    @NotNull
    @Valid
    private Map<String, Object> taxCriteria;

    @NotNull
    @Positive
    private BigDecimal total;

    private BigDecimal totalTaxes;

    private BigDecimal totalDiscount;

    @Valid
    private ThirdPartyDTO thirdParty;

    @Builder.Default
    private BigDecimal co2Total = BigDecimal.ZERO;

    private String uid;

    private UUID uuid;
}

File: projects/posv2/core/src/main/kotlin/com/robinfood/core/dtos/transactionrequest/OrderDTO.kt
package com.robinfood.core.dtos.transactionrequest

import com.robinfood.core.dtos.OrderThirdPartyDTO
import com.robinfood.core.dtos.OrderThirdPartyRequestDTO
import java.math.BigDecimal
import javax.validation.Valid
import javax.validation.constraints.Min
import javax.validation.constraints.NotBlank
import javax.validation.constraints.NotNull
import javax.validation.constraints.Positive
import javax.validation.constraints.Size

data class OrderDTO(

        @Positive
        val billingResolutionId: Long,

        @NotNull
        @Valid
        val brand: OrderBrandDTO,

        @NotNull
        val couponCriteriaInfo: Map<String, Any>,

        @NotNull
        @Size
        @Valid
        val coupons: List<CouponDTO>,

        @Min(0)
        val co2Total: BigDecimal?,

        @NotNull
        @Positive
        val deliveryTypeId: Long,

        @NotNull
        @Size
        @Valid
        val discounts: List<DiscountDTO>,

        @NotNull
        @Size(min = 1)
        @Valid
        val finalProducts: List<FinalProductDTO>,

        @Valid
        val flags: FlagsDTO?,

        @NotNull
        @Positive
        val flowId: Long,

        @Positive
        val id: Long? = null,

        @NotBlank
        val notes: String,

        @NotNull
        @Valid
        val origin: OriginDTO,

        @NotNull
        @Positive
        val paymentModelId: Long,

        @NotNull
        @Size
        @Valid
        val services: List<ServiceDTO>,

        @NotNull
        @Valid
        val store: StoreDTO,

        @NotNull
        val taxCriteria: Map<String, Any>,

        @Valid
        val thirdParty: OrderThirdPartyRequestDTO?,

        @NotNull
        @Positive
        val total: BigDecimal,

        var uuid: String?
)
